Add UrlAssetAccessor based on libcurl

kring opened the pull request:
This code was originally written by @timoore for [vsgCs](https://github.com/timoore/vsgCs), and is arriving here by way of an internal project.

I debated where to put it. The obvious choice is CesiumAsync, but that adds a large new dependency (libcurl) to a very widely used library. I decided in the end of put it in its own library (CesiumCurl).

timoore commented:
Good stuff. 

Current vsgCs adds the X-Cesium-Client headers to each request, in the same way as Cesium Unreal's `UnrealAssetAccessor`. I don't know if that's appropriate for a general asset accessor like this PR, but this functionality is a bit awkward for a client to add otherwise. One could subclass `CurlAssetAccessor`and wrap `get()` and `request()` to add the client headers to each request, but with the current function signatures that means a lot of copying of headers.

Perhaps these extra header fields could be passed to the constructor?

kring commented:
Good idea @timoore. I've added an option to add certain headers to all requests. I also moved all of the constructor parameters into an "options" class.

timoore commented:
> Good idea @timoore. I've added an option to add certain headers to all requests. I also moved all of the constructor parameters into an "options" class.

Cool. I think initializing OpenSSL can be kept totally separate in an application.

timoore commented on CesiumCurl/include/CesiumCurl/CurlAssetAccessor.h:
```diff
@@ -0,0 +1,130 @@
+ * @brief An implementation of `IAssetAccessor` that can make network and local
+ * requests to a variety of servers using libcurl.
+ */
+class CESIUMCURL_API CurlAssetAccessor final
```
Why `final`? I would say that we are putting this out there for users to customize.

kring commented on CesiumCurl/src/CurlAssetAccessor.cpp:
```diff
@@ -0,0 +1,528 @@
+    const std::string& url,
+    const HttpHeaders& headers,
+    const std::string& userAgent,
+    const std::string& certificatePath,
```
> On Linux, the key to smooth sailing with OpenSSL is to use the configuration file that goes with the library used.

This makes me nervous. If I understand correctly, it means that the set of certificate authorities will be baked into the application at compile time, and can't (easily?) be updated.

Using the operating system's certificate configuration seems like a much better idea. I'd be interested to learn what about Fedora's OpenSSL configuration breaks the vcpkg OpenSSL. Using the system OpenSSL makes sense where it's an option.

In any case, I don't think we're making any strong commitments here in this PR, just giving users options.

timoore replied:
> Using the operating system's certificate configuration seems like a much better idea. I'd be interested to learn what about Fedora's OpenSSL configuration breaks the vcpkg OpenSSL. Using the system OpenSSL makes sense where it's an option.

It's the option `rh-allow-sha1-signatures`. Not anything to do with certificates, but a random option that causes vcpkg to give up when it doesn't recognize it in opensslcnf.config... or at least what I could tell from debugging.

> In any case, I don't think we're making any strong commitments here in this PR, just giving users options.

timoore commented on CesiumCurl/include/CesiumCurl/CurlAssetAccessor.h:
```diff
@@ -0,0 +1,93 @@
+ * requests to a variety of servers using libcurl.
+ */
+class CESIUMCURL_API CurlAssetAccessor final
+    : public std::enable_shared_from_this<CurlAssetAccessor>,
```
OK, I see now. The original `UrlAssetAccessor` doesn't do this, relying unsafely on the acessor being alive for the duration of the application.
